lua/ttt2/libraries/gameloop.lua

```diff
@@ -0,0 +1,873 @@
+
+            -- this is a "fake" time shown to innocents, showing the end time if no
+            -- one would have been killed, it has no gameplay effect
+            gameloop.SetHasteEnd(timeRoundEnd)
+        end
+
+        gameloop.SetPhaseEnd(timeRoundEnd)
+        gameloop.StartWinChecks()
+        gameloop.StartNameChangeChecks()
+
+        ServerLog("Round proper has begun...\n")
+
+        ---
+        -- @realm server
+        -- stylua: ignore
+        hook.Run("TTT2PreBeginRound")
+
+        ---
+        -- @realm server
+        -- stylua: ignore
+        hook.Run("TTTBeginRound")
+    end
+
+    ---
+    -- Starts the wnd phase. Handles all setup needed to end a round.
+    -- @param string|number result The winning team or the win type
+    -- @internal
+    -- @realm server
+    function gameloop.End(result)
+        KARMA.RoundEnd()
+
+        -- update the round state that is used to determine how the game behaves
+        gameloop.SetRoundState(ROUND_POST)
+
+        -- triggers the round state output for every map setting entity
+        ents.TTT.TriggerRoundStateOutputs(ROUND_POST, result)
+
+        -- print the round result to the console
+        PrintResultMessage(result)
+
+        local timeEndPhase = math.max(5, cvDurationEndPhase:GetInt())
+
+        timer.Create("end2prep", timeEndPhase, 1, gameloop.Post)
+
+        gameloop.SetPhaseEnd(CurTime() + timeEndPhase)
+        gameloop.StopWinChecks()
+
+        -- send each client the role setup, reveal every player
+        local rlsList = roles.GetList()
+
+        for i = 1, #rlsList do
+            SendSubRoleList(rlsList[i].index)
+        end
+
+        -- We may need to start a timer for a mapswitch, or start a vote
+        gameloop.CheckForMapSwitch()
+
+        ---
+        -- @realm server
+        -- stylua: ignore
+        hook.Run("TTT2PreEndRound", result, timeEndPhase)
+
+        ---
+        -- @realm server
+        -- stylua: ignore
+        hook.Run("TTTEndRound", result)
+    end
+
+    ---
+    -- Is called after round end and cleans up the current round, calls the map cleanup and
+    -- therefore prepares the start of the next round.
+    -- @param[opt] boolean keepRoundCount Set to true to prevent the round count decrease
+    -- @internal
+    -- @realm server
+    function gameloop.Post(keepRoundCount)
+        loadingscreen.Begin()
+
+        if not keepRoundCount then
+            gameloop.DecreaseRoundsLeft()
+        end
+
+        -- delay the cleanup a bit so that the client starts the loading screen animation before the
+        -- cleanup starts
+        -- note: delaying by a single tick seemed to be ineffective
+        timer.Simple(0.25, function()
+            game.CleanUpMap(false, nil, ents.TTT.FixParentedPostCleanup)
+        end)
+
+        -- Remove ULX /me command. (the /me command is the only thing this hook does)
+        hook.Remove("PlayerSay", "ULXMeCheck")
```

Comment:
Do we know why we do this?
Is this a gameplay consideration or is this an "optimization" simply decreasing the amount of hooks?

Reply:
I've ggogled for this, because I was puzzled as well. Aparantly ULX adds this hook and it breaks gamemodes such as TTT and DarkRP. Therefore both gamemodes remove this hook again.

Seems really out of place imho, but I didn't want to add new issues, so I just kept it